File: Java/Tree/LowestCommonAncestorOfBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
	// the solution to count the number of targets in each subtree, 
	// and then compare and then use TreeNode 
	// time complexity is O(n^2) since for every node, almost, we need to count again 
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root==null) return root;
        if(countNode(root.left)==1 && countNode(root.right)==1) return root;
        if(countNode(root.left)==2) return lowestCommonAncestor(root.left, p, q);
        if(countNode(root.right)==2) return lowestCommonAncestor(root.right, p, q);
        if((root==q || root==p) && (counNode(root.left)==1 || countNode(root.right)==1)) 
        	return root;
    }

    private int countNode(TreeNode root, TreeNode p, TreeNode q){
    	if(root!=p && root!=q) return 0;
    	if(root==p || root==q) {
    		return countNode(root.left) + countNode(root.right) + 1;
    	}
    	return counNode(root.left) + counNode(root.right);
    }

    // more concise 
    // time complexity is O(N) since it is a simple traversal 
    // the traversal is preorder , parents first and then two children 
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q){
    	if(root==null) return null;
    	if(root==p || root==q) return root;
    	TreeNode left = lowestCommonAncestor(root.left, p, q);
    	TreeNode right = lowestCommonAncestor(root.right, p, q);
    	return (left!=null && right!=null)? root : right==null? left : right;
    }

    // iterative solution 
    
}

File: CC/Trees and Graph/lowestcommonancestorofbinarytree.js
/* Definition of a Binary Tree */
function TreeNode(val){
	this.val = val;
	this.left = this.right = null;
}

// method 1
// count node number idea 
var commonLowestAncestor = function(root, p, q){
	if(root === null || root===p || root===q) return root;
	var leftNum = getNodeNumber(root.left, p, q);
	var rightNum = getNodeNumber(root.right, p, q);

	if(leftNum === 2){
		commonLowestAncestor(root.left, p, q);
	}else if(rightNum === 2){
		commonLowestAncestor(root.right, p, q);
	}else{
		return root;
	}
};

function getNodeNumber(node, p, q){
	if(node === null){ 
		return 0;
	}else if(node === p || node===q ){
		return 1 + getNodeNumber(node.left, p, q) + getNodeNumber(node.right, p, q);
	}else{
		return getNodeNumber(node.left, p, q) + getNodeNumber(node.right, p, q);
	} 
}

// method 2

var commonLowestAncestor = function(root, p, q){
	if(root===null || root===p || root===q) return root;

	// only return if we find a target node or we did not find anything 
	var left = commonLowestAncestor(root.left, p, q);
	var right = commonLowestAncestor(root.right, p, q);

	if(left!==null && right!==null) return root;
	return left===null? right : left;
};

// method 3
// iterative method 
// use a parent dictionary
var commonLowestAncestor2 = function(root, p, q){
	var parentMap = new Map();
	parentMap.set(root, null);
	var stack = [];

	stack.push(root);
	while(!parentMap.has(p) || !parentMap.has(q)){
		var cur = stack.pop();
		if(cur.left!==null){
			parentMap.set(cur.left, cur);
			stack.push(cur.left);
		}
		if(cur.right!==null){
			parentMap.set(cur.right, cur);
			stack.push(cur.right);
		}
	}

	var ancestorSet = new Set();
	while(p!==null){
		ancestor.add(p);
		p = parentMap.get(p);
	}

	while(!ancestorSet.has(q)){
		q = parentMap.get(q);
	}
}; // time complexity O(N)

// Conditions when we have pointer to parent in node definition 

function TreeNodeWithParent(val){
	this.val = val;
	this.left = this.right = this.parent = null;
}

// Same idea as above method 
var lowestCommonAncestor = function(root, p, q){
	if(root === null || p === null || q === null) return null;

	var parentSet = {};
	// get all ancestors of p 
	while(p!==null) {
		parentSet.add(p);
		p = p.parent;
	}

	while(!parentSet.has(q)) q = q.parent;
	return q;
}; // Time Complexity is O(d) d is the depth of lower node 
